### ISSUE TITLE
Flaky test test_rearrange

### ISSUE THREAD

**Issue by TomAugspurger** [dask/dask#5867]
```
=================================== FAILURES ===================================

________________________ test_rearrange[processes-disk] ________________________

shuffle = 'disk', scheduler = 'processes'

    @pytest.mark.parametrize("shuffle", ["tasks", "disk"])

    @pytest.mark.parametrize("scheduler", ["threads", "processes"])

    def test_rearrange(shuffle, scheduler):

        df = pd.DataFrame({"x": np.random.random(10)})

        ddf = dd.from_pandas(df, npartitions=4)

        ddf2 = ddf.assign(_partitions=ddf.x % 4)

    

        result = rearrange_by_column(ddf2, "_partitions", max_branch=32, shuffle=shuffle)

        assert result.npartitions == ddf.npartitions

        assert set(ddf.dask).issubset(result.dask)

    

        # Every value in exactly one partition

>       a = result.compute(scheduler=scheduler)

dask/dataframe/tests/test_shuffle.py:272: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

dask/base.py:166: in compute

    (result,) = compute(self, traverse=False, **kwargs)

dask/base.py:437: in compute

    results = schedule(dsk, keys, **kwargs)

dask/multiprocessing.py:212: in get

    result = get_async(

dask/local.py:486: in get_async

    raise_exception(exc, tb)

dask/local.py:316: in reraise

    raise exc

dask/local.py:222: in execute_task

    result = _execute_task(task, data)

dask/core.py:119: in _execute_task

    return func(*args2)

dask/dataframe/shuffle.py:669: in shuffle_group_3

    p.append(d, fsync=True)

../../../miniconda/envs/test-environment/lib/python3.8/site-packages/partd/encode.py:25: in append

    self.partd.append(data, **kwargs)

../../../miniconda/envs/test-environment/lib/python3.8/site-packages/partd/file.py:37: in append

    if lock: self.lock.acquire()

../../../miniconda/envs/test-environment/lib/python3.8/site-packages/locket/__init__.py:99: in acquire

    lock.acquire()

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   self._file = open(self._path, "w")

E   FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpn5wlil7v.partd/.lock'
```

Will look into this a bit. Both were on 3.8 IIRC.

**Comment by TomAugspurger** [comment]
This seems to occur when the directory is removed prior between creating the `partd.File` object and attempting to append data to it. Roughly

```pytb
In [13]: import partd
In [14]: f = partd.File("data", )
In [16]: !rm -rf data

In [17]: f.append({"x": b"abc"})

---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-17-40cdad36fb6c> in <module>
----> 1 f.append({"x": b"abc"})

~/sandbox/partd/partd/file.py in append(self, data, lock, fsync, **kwargs)
     38     def append(self, data, lock=True, fsync=False, **kwargs):
     39         print(os.path.exists(self.filename('')))
---> 40         if lock: self.lock.acquire()
     41         try:
     42             for k, v in data.items():

~/sandbox/locket.py/locket/__init__.py in acquire(self)
    168     def acquire(self):
    169         self._lock.acquire(self._timeout, self._retry_period)
--> 170
    171     def release(self):
    172         self._lock.release()

~/sandbox/locket.py/locket/__init__.py in acquire(self, timeout, retry_period)
     98             for lock in self._locks:
     99                 lock.acquire(timeout, retry_period)
--> 100                 acquired_locks.append(lock)
    101         except:
    102             for acquired_lock in reversed(acquired_locks):

~/sandbox/locket.py/locket/__init__.py in acquire(self, timeout, retry_period)
    140         if self._file is None:
    141             self._file = open(self._path, "wb")
--> 142         if timeout is None and _lock_file_blocking_available:
    143             _lock_file_blocking(self._file)
    144         else:

FileNotFoundError: [Errno 2] No such file or directory: 'data/.lock'
```

It's not (yet) clear to my why that directory is being removed.

**Comment by TomAugspurger** [comment]
OK, I think what's happening is

1. Process A creates a `part.File` with no path. Partd makes a tmpdir, and marks it for removal when the `partd.File` is GCed
2. Process B reuses process A's partd.File's tmpdir for its own `partd.File`.
3. Process A's partd.File is GCed, triggering A's `part.File.drop`. This clears the tmpdir and removes it.
4. We have a race condition: If process B attempts to append data prior to A's `File.drop` being done, we get the exception. You can simulate this by inserting a `time.sleep` in `partd.File.drop`.

I don't know how to fix this yet.

**Comment by mrocklin** [comment]
So why would process B be appending data before A's drop is called?  I would only expect any process to clear out the data directory after all of them are done operating on it.

**Comment by TomAugspurger** [comment]
That I don't know. Here's some debug output. These are printing either the task key & PID in `dask.local.execute_task`, or the `file` passed to `partd.File` and the PID.

```
** task ('zpartd-4320f2f849cb11eab8c6186590cd1c87',) 24524
** file None /var/folders/hz/f43khqfn7b1b1g8z_z6y3bsw0000gp/T/tmp46ml086h.partd 24524
task ('getitem-mod-assign-04375fd34e33c3cc6fbc296cec0cdf70', 0) 24523
task ('getitem-mod-assign-04375fd34e33c3cc6fbc296cec0cdf70', 1) 24525
task ('getitem-mod-assign-04375fd34e33c3cc6fbc296cec0cdf70', 2) 24526
file /var/folders/hz/f43khqfn7b1b1g8z_z6y3bsw0000gp/T/tmp46ml086h.partd /var/folders/hz/f43khqfn7b1b1g8z_z6y3bsw0000gp/T/tmp46ml086h.partd 24471
task ('assign-04375fd34e33c3cc6fbc296cec0cdf70', 1) 24527
task ('assign-04375fd34e33c3cc6fbc296cec0cdf70', 2) 24528
task ('assign-04375fd34e33c3cc6fbc296cec0cdf70', 0) 24529
file /var/folders/hz/f43khqfn7b1b1g8z_z6y3bsw0000gp/T/tmp46ml086h.partd /var/folders/hz/f43khqfn7b1b1g8z_z6y3bsw0000gp/T/tmp46ml086h.partd 24525
task ('shuffle-partition-4320f2f849cb11eab8c6186590cd1c87', 0) 24525
file /var/folders/hz/f43khqfn7b1b1g8z_z6y3bsw0000gp/T/tmp46ml086h.partd /var/folders/hz/f43khqfn7b1b1g8z_z6y3bsw0000gp/T/tmp46ml086h.partd 24524
** task ('shuffle-partition-4320f2f849cb11eab8c6186590cd1c87', 1) 24524
file /var/folders/hz/f43khqfn7b1b1g8z_z6y3bsw0000gp/T/tmp46ml086h.partd /var/folders/hz/f43khqfn7b1b1g8z_z6y3bsw0000gp/T/tmp46ml086h.partd 24530
task ('shuffle-partition-4320f2f849cb11eab8c6186590cd1c87', 2) 24530
partd.File.append          24525
partd.File.drop-start      24524
partd.File.append          24530
(Pdb) partd.File.drop-done 24524
partd.File.append          24524
```

I've prefixed some lines with `**`.

Process `24524` created the initial `partd.File` without a path during `zpartd-...`, so partd made it and marked it for deletion when that `File` goes out of scope. Other processes start using that directory. Then process `24524` moved on to `shuffle-partition-...`, which (I think?) will cause the `File` to go out of scope and eventually get GCed. That's logged with a `partd.File.drop-start`, and finishes with a `partd.File.drop-done` (I added a 3 second `time.sleep` in there). In the meantime, a few other processes did appends to that directory.

**Comment by TomAugspurger** [comment]
So, IMO, the easiest fix here is to have the initial task creating the file pass a Path (which disables partd.File's auto-cleanup), and *perhaps* add a task at the end that cleans stuff up?

**Comment by mrocklin** [comment]
Hrm, but if we fail partway through the computation then we won't clean things up?

cc also @jcrist , who I think had a hand in the multi-processing Partd implementation (although no pressure)

**Comment by mrocklin** [comment]
@jcrist I think that you volunteered to take a look at this.  Is this still something that you can do?

**Comment by jcrist** [comment]
I dove into this this morning, and wasn't able to reproduce on either 3.7 or 3.8. Tom eventually was able to reproduce locally, but only after a bit of work, running the same reproducer on my machine doesn't trigger anything even after 500 runs. I'm not sure what's going on here, but without being able to reproduce the issue I'm not sure what I can do to help.

**Comment by KrishanBhasin** [comment]
I'm getting this error when running compute fairly consistently at the moment.

I'm currently playing around with my worker/thread distribution on my LocalCluster, and this mainly seems to crop up when I have higher process count vs lower (8 vs 2)

**Comment by TomAugspurger** [comment]
@KrishanBhasin are you interested in working on a fix? My best understanding is in https://github.com/dask/dask/issues/5867#issuecomment-583479468, but I haven't come up with an actual solution yet.

**Comment by KrishanBhasin** [comment]
I'd love to but I doubt I have the required knowledge to take a good stab at this.

**Comment by KrishanBhasin** [comment]
It looks like I get this error pretty consistently when I run a job on a LocalCluster that has been up for a bit, and had workers die/be killed previously.

I had to kill off some workers off mid-compute, and later in the compute got the distributed error : https://github.com/dask/distributed/pull/3575 (with no results written out; the job was meant to end with a `to_csv()` 😢 )

I reused the cluster & the running workers to rerun the compute, and ran into this bug a few times in a row.

**Comment by TomAugspurger** [comment]
https://github.com/dask/dask/pull/5977 is getting closer, but I've been a bit short on time this past week.

**Comment by KrishanBhasin** [comment]
That’s great to hear! I didn’t mean to sound like I was chasing, I was just hoping extra context might help in hunting it down the source of the bug.
